kroxylicious-filters/kroxylicious-encryption/doc/design.adoc

```diff
@@ -0,0 +1,267 @@
+----
+record                     = length
+                             attributes
+                             timestampDelta
+                             offsetDelta
+                             recordKeyLength
+                             recordKey
+                             recordValueLen
+                             recordValue
+                             *recordHeader
+recordHeader               = headerKeyLength
+                             headerKey
+                             headerValueLength
+                             headerValue
+----
+
+To construct the "ciphertext record":
+
+. Construct the `parcel` from the plaintext record's `value` and optionally the `headers`
+. Encrypt the `parcel`, and construct the `wrapping`
+. Create a copy of the plaintext record, but:
+.. replace the `value` with the `wrapping` (adjusting the `valueLen`)
+.. optionally clear the `headers` (if they were included in the parcel)
+.. Prepend a header with name "kroxylicious.io/encryption" and byte value `0x01`.
+
+
+=== Compression
+
+Kafka supports compression of record batches (not individual records). This means there are a number of places where compression could be used:
+
+==== Client sends a compressed batch
+It's _batches_ that are compressed (by the producer), but _records_ that are encrypted, so we have no choice but to decompress the whole batch if we're to encrypt any of the records within it.
+There is the possibility of (re-)compressing records prior to encryption.
+Compress-then-encrypt has been used for cryptanalysis (e.g. the CRIME vulnerability in HTTP). So we don't currently do this.
+
+
+==== Proxy compresses a batch of encrypted records, prior to forwarding to a broker
+
+It's the proxy's choice about which compression to use. In other words whatever compressor is used a proxied client won't be exposed to the compressed batches.
+
+Since compression is a trade-off between CPU and compression ratio, we should offer compression as an configurable option.
+
+==== Proxy decompresses a batch of records, when receiving from a broker
+
+The proxy is stateless. In particular the proxy does not maintain state about which topics have had encrypted records produced to them. This means there is no state to manage, and users can easily change which records are encrypted, and how, with confidence that existing records will still be readable.
+
+However, as consequence of this design, on the fetch path we have no choice but to decompress batches in order to determine whether they contain encrypted records. Encrypted records are identified by the presence of the "kroxylicious.io/encryption" header.
+
+==== Proxy compresses a batch of decrypted records, prior to forwarding to the client
+
+The benefits of doing this are unclear. The compression only benefits a single client and only in terms of network bandwidth. It's unclear whether it's worth the CPU. Likely we should only support a single compressor, chosen based on experiment.
```

Comment:
You don't think there is an argument for restoring the publisher's chosen compression?  In my mind for the proxy to be truly transparent it should at least have the option to behave in this way.